flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestServerEndpoint.java

```diff
@@ -364,22 +364,37 @@ public String getRestBaseUrl() {
 		}
 	}
 
-	private static void registerHandler(Router router, Tuple2<RestHandlerSpecification, ChannelInboundHandler> specificationHandler) {
-		switch (specificationHandler.f0.getHttpMethod()) {
+	private static void registerHandler(Router router, Tuple2<RestHandlerSpecification, ChannelInboundHandler> specificationHandler, Logger log) {
+		final String handlerURL = specificationHandler.f0.getTargetRestEndpointURL();
+		// setup versioned urls
+		for (final RestAPIVersion supportedVersion : specificationHandler.f0.getSupportedAPIVersions()) {
+			final String versionedHandlerURL = '/' + supportedVersion.getURLVersionPrefix() + handlerURL;
+			log.debug("Register handler {} under {}@{}.", specificationHandler.f1, specificationHandler.f0.getHttpMethod(), versionedHandlerURL);
+			registerHandler(router, versionedHandlerURL, specificationHandler.f0.getHttpMethod(), specificationHandler.f1);
+		}
+		// setup unversioned url for convenience and backwards compatibility
+		// this url will always point to the oldest supported version
+		// this is controlled by the order in which handler are registered
+		log.debug("Register handler {} under {}@{}.", specificationHandler.f1, specificationHandler.f0.getHttpMethod(), handlerURL);
+		registerHandler(router, handlerURL, specificationHandler.f0.getHttpMethod(), specificationHandler.f1);
```

Comment:
Shouldn't we explicitly register the legacy handlers? I think it might be a bit confusing if we change a handler which no longer supports the legacy format and still register it here. We could, for example introduce a legacy version `RestAPIVersion.Unversioned` which we register without the version prefix.

Reply:
I'm not sure what you mean with legacy handlers.

The goal here is to always keep the oldest version as the default. So, currently `v1` is the default, and if we remove that at any point then `v2` will be the new default.

Basically, we have 2 options for dealing with unversioned URLS in the future:
* remove unversioned URLs altogether, in which case explicit UNVERSIONED is useful for the time being
* always keep a default version, in which case UNVERSIONED is not useful as we have to carry it over whenever we remove an API which is prone to errors

Reply:
Ah ok, the idea is that the unversioned URLs point to the oldest supported REST API version, right? The problem with this approach is that we would break the API (unversioned URLs) every time we drop support for a given version. Given that the versioned URLs are stable and we did not promise REST API stability yet, this should most likely be ok. 

The breaking change would only happen if users of the unversioned URLs don't migrate to the versioned URLs and we drop support for a version. Once we do this, we should add it to the migration guide.

Could we maybe add a check which only registers handlers of the oldest version under the unversioned URLs instead of all?

Reply:
I just realized that the current approach in the PR has a troublesome flaw: the default API is actually mixed iff a newer version defines a new url. (if we add `/totally/new/url` in `v2` then the default for other urls would be `v1`, but for the new one it would be `v2`). 
As you suggested we should restrict the default to a specific version, which would also solve the issue of multiple registrations.

What we could do is add a `isDefault` flag to the `RestAPIVersion` enum, and only register unversioned urls for that version. We could enforce with a test that exactly one version is marked as the default. WDYT?

Reply:
Sounds like a good idea to me. Good catch @zentol!